ingest/downloader/flattener.py

```diff
@@ -0,0 +1,109 @@
+from typing import List
+
+ONTOLOGY_REQUIRED_PROPS = ['ontology', 'ontology_label']
+EXCLUDE_KEYS = ['describedBy', 'schema_type']
+
+
+class Flattener:
+    def __init__(self):
+        self.workbook = {}
+
+    def flatten(self, entity_list: List[dict], object_key: str = ''):
+        for entity in entity_list:
+            self._flatten_entity(entity, object_key)
+        return self.workbook
+
+    def _flatten_entity(self, entity, object_key):
+        worksheet_name = object_key
+        row = {}
+        content = entity
+
+        if not object_key:
+            content = entity['content']
+            worksheet_name = self._get_concrete_entity(content)
+            row = {f'{worksheet_name}.uuid': entity['uuid']['uuid']}
+
+        if not worksheet_name:
+            raise Exception('There should be a worksheet name')
+
+        self._flatten_object(content, row, parent_key=worksheet_name)
+
+        user_friendly_worksheet_name = self._format_worksheet_name(worksheet_name)
+        worksheet = self.workbook.get(user_friendly_worksheet_name, {'headers': [], 'values': []})
+
+        rows = self._update_rows(row, worksheet)
+        headers = self._update_headers(row, worksheet)
+
+        self.workbook[user_friendly_worksheet_name] = {
+            'headers': headers,
+            'values': rows
+        }
+
+    def _update_rows(self, row, worksheet):
+        rows = worksheet.get('values')
+        rows.append(row)
+        return rows
+
+    def _update_headers(self, row, worksheet):
+        headers = worksheet.get('headers')
+        for key in row.keys():
+            if key not in headers:
+                headers.append(key)
+        return headers
+
+    def _flatten_object(self, object: dict, flattened_object: dict, parent_key: str = ''):
+        if isinstance(object, dict):
+            for key in object:
+                if key in EXCLUDE_KEYS:
+                    continue
+
+                value = object[key]
+                full_key = f'{parent_key}.{key}' if parent_key else key
+                if isinstance(value, dict) or isinstance(value, list):
+                    self._flatten_object(value, flattened_object, parent_key=full_key)
+                else:
+                    flattened_object[full_key] = str(value)
+        elif isinstance(object, list):
+            self._flatten_list(flattened_object, object, parent_key)
+
+    def _flatten_list(self, flattened_object, object, parent_key):
+        if self._is_list_of_objects(object):
+            self._flatten_object_list(flattened_object, object, parent_key)
+        else:
+            self._flatten_scalar_list(flattened_object, object, parent_key)
+
+    def _flatten_scalar_list(self, flattened_object, object, parent_key):
+        stringified = [str(e) for e in object]
+        flattened_object[parent_key] = '||'.join(stringified)
```

Comment:
i would make all the delimiters used in this file constants. `||`, `-`, ` - `...

Reply:
Done.